src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,2233 @@
+  from void* pointer.  If either is missing it returns ARK_MEM_NULL.
+  ---------------------------------------------------------------*/
+int lsrkStep_AccessARKODEStepMem(void* arkode_mem, const char* fname,
+                                 ARKodeMem* ark_mem, ARKodeLSRKStepMem* step_mem)
+{
+  /* access ARKodeMem structure */
+  if (arkode_mem == NULL)
+  {
+    arkProcessError(NULL, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_ARK_NO_MEM);
+    return ARK_MEM_NULL;
+  }
+  *ark_mem = (ARKodeMem)arkode_mem;
+
+  /* access ARKodeLSRKStepMem structure */
+  if ((*ark_mem)->step_mem == NULL)
+  {
+    arkProcessError(*ark_mem, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_LSRKSTEP_NO_MEM);
+    return ARK_MEM_NULL;
+  }
+  *step_mem = (ARKodeLSRKStepMem)(*ark_mem)->step_mem;
+  return ARK_SUCCESS;
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_AccessStepMem:
+
+  Shortcut routine to unpack the step_mem structure from
+  ark_mem.  If missing it returns ARK_MEM_NULL.
+  ---------------------------------------------------------------*/
+int lsrkStep_AccessStepMem(ARKodeMem ark_mem, const char* fname,
+                           ARKodeLSRKStepMem* step_mem)
+{
+  /* access ARKodeLSRKStepMem structure */
+  if (ark_mem->step_mem == NULL)
+  {
+    arkProcessError(ark_mem, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_LSRKSTEP_NO_MEM);
+    return ARK_MEM_NULL;
+  }
+  *step_mem = (ARKodeLSRKStepMem)ark_mem->step_mem;
+  return ARK_SUCCESS;
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_CheckNVector:
+
+  This routine checks if all required vector operations are
+  present.  If any of them is missing it returns SUNFALSE.
+  ---------------------------------------------------------------*/
+sunbooleantype lsrkStep_CheckNVector(N_Vector tmpl)
+{
+  if ((tmpl->ops->nvclone == NULL) || (tmpl->ops->nvdestroy == NULL) ||
+      (tmpl->ops->nvlinearsum == NULL) || (tmpl->ops->nvconst == NULL) ||
+      (tmpl->ops->nvscale == NULL) || (tmpl->ops->nvwrmsnorm == NULL) ||
+      (tmpl->ops->nvspace == NULL))
+  {
+    return SUNFALSE;
+  }
+  return SUNTRUE;
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_DomEigUpdateLogic:
+
+  This routine checks if the step is accepted or not and reassigns
+  the dom_eig update flags accordingly.
+  ---------------------------------------------------------------*/
+
+void lsrkStep_DomEigUpdateLogic(ARKodeMem ark_mem, ARKodeLSRKStepMem step_mem,
+                                sunrealtype dsm)
+{
+  if (dsm <= ONE)
+  {
+    N_VScale(ONE, ark_mem->tempv2, ark_mem->fn);
+    ark_mem->fn_is_current = SUNTRUE;
+
+    step_mem->dom_eig_is_current = (step_mem->const_Jac == SUNTRUE);
+
+    step_mem->dom_eig_update = SUNFALSE;
+    if (step_mem->dom_eig_nst + step_mem->dom_eig_freq <=
+        ark_mem->nst_attempts + 1)
```

Comment:
This should be based on `nst` not `nst_attempts`
```suggestion
    if (ark_mem->nst + 1 >= step_mem->dom_eig_nst + step_mem->dom_eig_freq)
```